evap/staff/importers.py

```diff
@@ -97,28 +98,99 @@ def store_in_database(self, vote_start_datetime, vote_end_date, semester):
         assert not self.errors
         # This is safe because the user's email address is checked before in the importer (see #953)
         responsible_dbobj = UserProfile.objects.get(email=self.responsible_email)
-        course = Course(
-            name_de=self.name_de,
-            name_en=self.name_en,
-            type=self.course_type,
+        if self.existing_course:
+            self.existing_course.degrees.add(*self.degrees)
+        else:
+            course = Course(
+                name_de=self.name_de,
+                name_en=self.name_en,
+                type=self.course_type,
+                semester=semester,
+            )
+            course.save()
+            course.responsibles.set([responsible_dbobj])
+            course.degrees.set(self.degrees)
+            evaluation = Evaluation(
+                vote_start_datetime=vote_start_datetime,
+                vote_end_date=vote_end_date,
+                course=course,
+                wait_for_grade_upload_before_publishing=self.is_graded,
+            )
+            evaluation.save()
+            evaluation.contributions.create(
+                evaluation=evaluation,
+                contributor=responsible_dbobj,
+                role=Contribution.Role.EDITOR,
+                textanswer_visibility=Contribution.TextAnswerVisibility.GENERAL_TEXTANSWERS,
+            )
+
+    def check_existing_course(self, semester, importer):
+        name_collision = self.check_name_collision(semester)
+        if not name_collision:
+            return False
+
+        self.find_and_set_existing_course(semester, importer)
+        if self.existing_course is None:
+            if name_collision == "both":
+                importer.errors[ImporterError.COURSE].append(
+                    _("Course {} ({}) does already exist in this semester.").format(self.name_en, self.name_de)
+                )
+            else:
+                importer.errors[ImporterError.COURSE].append(
+                    _("Course {} does already exist in this semester.").format(getattr(self, name_collision))
+                )
```

Comment:
For the sanity of the reader, I think it would make sense to create these errors where the other errors are created. Currently, following the logic of this is quite the hopping between this, `check_name_collision` and `find_and_set_existing_course`, and you have to keep 6 combinations of `name_collision` and `existing_course` in mind while doing so.

Maybe one more inversion of the flow could help with that, too:
- If there exists a course with the same name_de and name_en, handle it (add mismatch error or set existing_course) and return
- (Implicit else) If there exists a course with either the same name_de or the same name_en, handle it (add errors I think)

Reply:
I just had another thought. Why do we display different error messages if there is already a name mismatch (either name_de or name_en does not match). We could also integrate those attributes in this error message:
```
importer.errors[ImporterError.COURSE].append(
    _("Course {} ({}) does already exist in this semester but the {} {} not match.").format(
        self.name_en,
        self.name_de,
        " and ".join(differing_attributes),
        ngettext("does", "do", len(differing_attributes)),
    )
)
```
So the result would be: There is one error message that lists all differing attributes of a course, if one of name_de or name_en matches but the other does not, or if both are equal and any other attribute is different.
Would that be a viable solution?

Reply:
In general: I think handling all the attributes equally makes sense, especially if it simplifies stuff.

The error message you listed still has the translation problem (https://github.com/e-valuation/EvaP/pull/1683#discussion_r796899279).

Reply:
Ok, I tried to move the generation of error message into one method and also have that early check for the name_collision you suggested in your first comment (I did it the other way around, but should be the same). See https://github.com/e-valuation/EvaP/pull/1683/commits/5897433102b11c9757214fdc507fb132794e35a0.
So now there is a method that checks a given course for its attributes (`check_existing_course_attributes`) whether they match with the current evaluationData instance. Then it returns all mismatching attributes and whether the number of evaluations is correct. These information then get transformed into an error message in the calling method. (Little side note: in `check_existing_course_attributes` there is an early return for a mismatching evaluation count because if there was no early return, the check for the grading property would maybe give false information or would crash)
Does the new structure improve understandability?

Would you still like to see what I suggested here: https://github.com/e-valuation/EvaP/pull/1683#discussion_r800857244? Because I am not sure if it makes the code even more complicated.

Reply:
> Would you still like to see what I suggested here: [#1683 (comment)](https://github.com/e-valuation/EvaP/pull/1683#discussion_r800857244)? Because I am not sure if it makes the code even more complicated.

I think the current code in the PR is fine. imho the aggregation you proposed above does not work well because there could be multiple merge candidates then, and then the messages would become misleading: The candidates would likely be two different courses, and obviously the user would not want to merge with both of them.

Resolving for now.